server/src/test/java/org/elasticsearch/test/BackgroundIndexer.java

```diff
@@ -0,0 +1,322 @@
+
+    private final Logger logger = LogManager.getLogger(getClass());
+
+    final Thread[] writers;
+    final SQLTransportExecutor sqlExecutor;
+    final CountDownLatch stopLatch;
+    final Collection<Exception> failures = new ArrayList<>();
+    final AtomicBoolean stop = new AtomicBoolean(false);
+    final AtomicLong idGenerator = new AtomicLong();
+    final CountDownLatch startLatch = new CountDownLatch(1);
+    final AtomicBoolean hasBudget = new AtomicBoolean(false); // when set to true, writers will acquire writes from a semaphore
+    final Semaphore availableBudget = new Semaphore(0);
+    private final Set<String> ids = ConcurrentCollections.newConcurrentSet();
+    private volatile Consumer<Exception> failureAssertion = null;
+
+    volatile int minFieldSize = 10;
+    volatile int maxFieldSize = 140;
+
+    /**
+     * Start indexing in the background using a random number of threads. Indexing will be paused after numOfDocs docs has
+     * been indexed.
+     *
+     * @param table     table name to write data
+     * @param column    column name to write data
+     * @param client    client to use
+     * @param numOfDocs number of document to index before pausing. Set to -1 to have no limit.
+     */
+    public BackgroundIndexer(String table, String column, SQLTransportExecutor client, int numOfDocs) {
+        this(table, column, client, numOfDocs, RandomizedTest.scaledRandomIntBetween(2, 5));
+    }
+
+    /**
+     * Start indexing in the background using a given number of threads. Indexing will be paused after numOfDocs docs has
+     * been indexed.
+     *
+     * @param table       table name to write data
+     * @param column      column name to write data
+     * @param client      client to use
+     * @param numOfDocs   number of document to index before pausing. Set to -1 to have no limit.
+     * @param writerCount number of indexing threads to use
+     */
+    public BackgroundIndexer(String table, String column, SQLTransportExecutor client, int numOfDocs, final int writerCount) {
+        this(table, column, client, numOfDocs, writerCount, true, null);
+    }
+
+    /**
+     * Start indexing in the background using a given number of threads. Indexing will be paused after numOfDocs docs has
+     * been indexed.
+     *
+     * @param table       table name to write data
+     * @param column      column name to write data
+     * @param numOfDocs   number of document to index before pausing. Set to -1 to have no limit.
+     * @param writerCount number of indexing threads to use
+     * @param autoStart   set to true to start indexing as soon as all threads have been created.
+     * @param random      random instance to use
+     */
+    public BackgroundIndexer(final String table, final String column, final SQLTransportExecutor sqlExecutor, final int numOfDocs, final int writerCount,
+                             boolean autoStart, Random random) {
+        if (random == null) {
```

Comment:
I removed `useAutoGeneratedIDs` because afaik primary keys can not be autogenerated in crate.

Reply:
If no primary key is defined on the table, the internal id `_id` will be generated, otherwise the values of the PK cols are used to compute a `_id` value.
As the `_id` column can be selected, I see no issue on keeping the `useAutogeneratedIDs` feature. Do I miss anything?

Reply:
This requires to initialise the table with or without primary key depending if the id is autogenerated or not. If the primary key is defined it cannot be null, so there is a dependency between the table definition and the indexing routine. However, i can do that.

Reply:
Ah right, so I missed something indeed ;-)
I've looked into this again and it seems that using autogenerated id's is not required for any current scenario.
I'd then prefer to not use it like you did.